mars/services/task/execution/ray/executor.py

```diff
@@ -235,32 +245,33 @@ async def execute_subtask_graph(
                 meta_object_ref, *output_object_refs = output_object_refs
                 # TODO(fyrestone): Fetch(not get) meta object here.
                 output_meta_object_refs.append(meta_object_ref)
-            context.update(zip(output_keys, output_object_refs))
+            task_context.update(zip(output_keys, output_object_refs))
         logger.info("Submitted %s subtasks of stage %s.", len(subtask_graph), stage_id)
 
         key_to_meta = {}
         if len(output_meta_object_refs) > 0:
             # TODO(fyrestone): Optimize update meta by fetching partial meta.
+            meta_count = len(output_meta_object_refs)
+            logger.info("Getting %s metas of stage %s.", meta_count, stage_id)
             meta_list = await asyncio.gather(*output_meta_object_refs)
             for meta in meta_list:
                 key_to_meta.update(meta)
             assert len(key_to_meta) == len(result_meta_keys)
-            logger.info(
-                "Got %s metas of stage %s.", len(output_meta_object_refs), stage_id
-            )
+            logger.info("Got %s metas of stage %s.", meta_count, stage_id)
 
         chunk_to_meta = {}
-        output_object_refs = []
+        # ray.wait requires the object ref list is unique.
+        output_object_refs = set()
         for chunk in chunk_graph.result_chunks:
             chunk_key = chunk.key
-            object_ref = context[chunk_key]
-            output_object_refs.append(object_ref)
+            object_ref = task_context[chunk_key]
+            output_object_refs.add(object_ref)
             chunk_meta = key_to_meta.get(chunk_key)
             if chunk_meta is not None:
                 chunk_to_meta[chunk] = ExecutionChunkResult(chunk_meta, object_ref)
 
         logger.info("Waiting for stage %s complete.", stage_id)
-        ray.wait(output_object_refs, fetch_local=False)
+        ray.wait(list(output_object_refs), fetch_local=False)
```

Comment:
Will `rat.wait` block event loop? Can we invoke it with `to_thread`?

Reply:
Good suggestion. Thanks.